### ISSUE TITLE
Support for cyclic dependency in models

### ISSUE THREAD

**Issue by MarinPostma** [chiselstrike/chiselstrike#1448]
#1443 fixed a bug in the handling of dependent models. In the doing, we decided to error when cycles were detected in model dependency because we are unable to handle them at the moment.

This issue discusses what would need to be changed to handle cyclic model dependency in the general case.

## Handling cyclic dependency at the type level

The first issue we'll encounter when dealing with cyclic dependency in models in building its `Type` instance (https://github.com/chiselstrike/chiselstrike/blob/main/server/src/types.rs#L487-L494) since `Type` is recursive on the `ObjectType`'s field. We'd need to refactor this code to have some kind of indirection on the  `Type` and avoid direct type recursion.

## Storing entities with cyclic dependencies

This should not be an issue in the relational model, we'd have each table have external references to the entity in the foreign table, no biggie.

## Retrieving entities with a cyclic dependency

To do that we'd need laziness or some other way (some kind of graph traversal?) to prevent infinite recursion when retrieving the object.

**Comment by espindola** [comment]
For retrieval,  we could create a cyclic object, no? If we have models A -> B -> C, one option is to always create a JS object for A, B and C, so they can refer to each other.

**Comment by MarinPostma** [comment]
This is the issue with eagerly loading, @BearLemma is working on collections, I will read his proposal, maybe there is something that could help us here.

**Comment by MarinPostma** [comment]
@espindola yes yes, but you would potentially be eagerly loading a long chain of those objects (e.g a chained list)

**Comment by espindola** [comment]
> @espindola yes yes, but you would potentially be eagerly loading a long chain of those objects (e.g a chained list)

Fair point. I *think* that is OK, since we do offer .select which can be used to fetch only a few fields, and should, module bugs, avoid reading fields that were not requested.

**Comment by MarinPostma** [comment]
If that ever happens, the workaround would be to wrap it in the lazy type @BearLemma is working on for one-to-many relationships

**Comment by 7flash** [comment]
That's biggest pain when working with Java JPA models..
